src/undo/LineStyleUndoAction.cpp

```diff
@@ -0,0 +1,86 @@
+#include "LineStyleUndoAction.h"
+
+#include "gui/Redrawable.h"
+#include "model/PageRef.h"
+#include "model/Stroke.h"
+
+#include "Rectangle.h"
+#include "i18n.h"
+
+class LineStyleUndoActionEntry {
+public:
+    LineStyleUndoActionEntry(Stroke* s, LineStyle oldStyle, LineStyle newStyle):
+            s(s), oldStyle(oldStyle), newStyle(newStyle) {}
+    Stroke* s;
+    LineStyle oldStyle;
+    LineStyle newStyle;
+};
+
+LineStyleUndoAction::LineStyleUndoAction(const PageRef& page, Layer* layer): UndoAction("LineStyleUndoAction") {
+    this->page = page;
+    this->layer = layer;
+}
+
+LineStyleUndoAction::~LineStyleUndoAction() {
+    for (LineStyleUndoActionEntry* e: this->data) {
+        delete e;
+    }
+}
+
+void LineStyleUndoAction::addStroke(Stroke* s, LineStyle originalStyle, LineStyle newStyle) {
+    this->data.push_back(new LineStyleUndoActionEntry(s, originalStyle, newStyle));
```

Comment:
Always use emplace over push / insert. Especially after changing `data`'s type to a vector holding structs.
```suggestion
    this->data.emplace_back(s, originalStyle, newStyle);
```